src/platform/posix/tcp_ip_channel.c

```diff
@@ -250,40 +289,27 @@ void TcpIpChannel_ctor(TcpIpChannel *self, const char *host, unsigned short port
     throw("Could not set SO_REUSEADDR");
   }
 
-  self->server = true;
+  // Set server socket to non-blocking
+  if (server) {
+    fcntl(self->fd, F_SETFL, O_NONBLOCK);
+  }
+
+  self->server = server;
   self->terminate = true;
   self->protocol_family = protocol_family;
   self->host = host;
   self->port = port;
   self->read_index = 0;
   self->client = 0;
 
-  self->super.accept = TcpIpChannel_accept;
-  self->super.bind = TcpIpChannel_bind;
-  self->super.connect = TcpIpChannel_connect;
-  self->super.close = TcpIpChannel_close;
-  self->super.receive = TcpIpChannel_receive;
-  self->super.send = TcpIpChannel_send;
-  self->super.register_callback = TcpIpChannel_register_callback;
+  self->super.open_connection = TcpIpChannel_open_connection;
+  self->super.try_connect = TcpIpChannel_try_connect;
+  self->super.close_connection = TcpIpChannel_close_connection;
+  self->super.send_blocking = TcpIpChannel_send_blocking;
+  self->super.register_receive_callback = TcpIpChannel_register_receive_callback;
   self->super.free = TcpIpChannel_free;
   self->receive_callback = NULL;
   self->federated_connection = NULL;
-}
-
-void TcpIpChannel_free(NetworkChannel *untyped_self) {
-  LF_DEBUG(NET, "Freeing TCP/IP Channel");
-  TcpIpChannel *self = (TcpIpChannel *)untyped_self;
-  self->terminate = true;
 
-  if (self->receive_thread != 0) {
-    LF_DEBUG(NET, "Stopping receive thread");
-    if (pthread_cancel(self->receive_thread) != 0) {
-      LF_ERR(NET, "Error canceling receive thread");
-    }
-
-    if (pthread_join(self->receive_thread, NULL) != 0) {
-      LF_ERR(NET, "Error joining receive thread");
-    }
-  }
-  self->super.close((NetworkChannel *)self);
+  self->receive_thread = 0;
```

Comment:
This is needed so that the `free` function does not cleanup threads that don't exist anymore if you clean and recreate the same tcpchannel. (I noticed this when running my new unit test)

Reply:
Because sometimes we recreate the tcp_ip_channel but without a new thread so the thread is still closed in reality